tools/fast_pusher_.py

```diff
@@ -98,28 +102,33 @@ def main():
   if not args.config and not args.tarball:
     raise Exception('Either --config or --tarball must be specified.')
 
+  if len(args.digest or []) != len(args.layer or []):
+    raise Exception('--digest and --layer must have matching lengths.')
+
   # If config is specified, use that.  Otherwise, fallback on reading
   # the config from the tarball.
   config = args.config
   if args.config:
+    logging.info('Reading config from %r', args.config)
     with open(args.config, 'r') as reader:
       config = reader.read()
   elif args.tarball:
+    logging.info('Reading config from tarball %r', args.tarball)
     with v2_2_image.FromTarball(args.tarball) as base:
       config = base.config_file()
 
-  if len(args.digest or []) != len(args.layer or []):
-    raise Exception('--digest and --layer must have matching lengths.')
-
-  transport = transport_pool.Http(httplib2.Http, size=_THREADS)
-
   # Resolve the appropriate credential to use based on the standard Docker
   # client logic.
+  logging.info('Loading Docker credentials for repository %r', str(name))
```

Comment:
Since you mentioned tracking where time was spent, you reminded me of some incomplete work someone contributed during a FixIt, which was aimed at creating a pluggable way of doing this.

The idea was that the various classes would take a constructor for a [`Context`](https://github.com/google/containerregistry/blob/master/client/monitor_.py#L47) implementation, and then wrap blocks of code in it.  So for example:
```python
def __init__(self, ..., context=monitor.Nop)
   ...
   self.context = context

def expensive_method(self):
   ...
   with self.context('expensive method block: foo'):
      # foo
   ...
```

This way clients wrapping the library could customize how these blocks are monitored.  `logging` is a trivial version of this, but I am very sure you can imagine more sophisticated implementations of `monitor.Context` that export to external metric systems :)